From af2cdb0f095f959594cfaeed5431bef9f2424389 Mon Sep 17 00:00:00 2001
From: bartimaeusnek <bartimaeusnek@noreply.codeberg.org>
Date: Thu, 22 Jan 2026 19:18:47 +0100
Subject: [PATCH 124/133] std.zig.system.loongarch: implement individual cpu
 feature bit tests (#30915)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
X-Developer-Signature: v=1; a=openpgp-sha256; l=2808; i=xtex@astrafall.org;
 h=from:subject; bh=I8mzp0cW+kW1QNY8jcqTcgzf+FxZhYbZwwncaxfcLU0=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78vgrzUWiFno80RX+5/ieJ3l/u2Kk0pmnJm0WM+IX9
 pxZtu54x0QWBjEuBksxRZYiwwZvVp10ftFl5bIwc1iZQIZIizQwMDAwsDDw5SbmlRrpGOmZahvq
 GRrqGOkYMXBxCsBUT1nJ8FfA1GhuZsysZK3MxXc3JQStSBDZEX3k+gK+U4Ln90inS+gw/JXes+1
 +nPexhiM9ii+iHvGYdBkv+J6uFOLysbNB7Sv7BiYA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

closes #30902

Co-authored-by: bartimaeusnek <33183715+bartimaeusnek@users.noreply.github.com>
Reviewed-on: https://codeberg.org/ziglang/zig/pulls/30915
Reviewed-by: Alex Rønne Petersen <alex@alexrp.com>
Co-authored-by: bartimaeusnek <bartimaeusnek@noreply.codeberg.org>
Co-committed-by: bartimaeusnek <bartimaeusnek@noreply.codeberg.org>
(cherry picked from commit 37288e53ae4e617a56fa79d5b718f3b66957d4f5)
---
 lib/std/zig/system/loongarch.zig | 38 +++++++++++++++++++++++++++++---
 1 file changed, 35 insertions(+), 3 deletions(-)

diff --git a/lib/std/zig/system/loongarch.zig b/lib/std/zig/system/loongarch.zig
index cdeeb10564a1..7a2c62279131 100644
--- a/lib/std/zig/system/loongarch.zig
+++ b/lib/std/zig/system/loongarch.zig
@@ -1,6 +1,16 @@
 const builtin = @import("builtin");
 const std = @import("std");
 
+inline fn bit(input: u32, offset: u5) bool {
+    return (input >> offset) & 1 != 0;
+}
+
+fn setFeature(cpu: *std.Target.Cpu, feature: std.Target.loongarch.Feature, enabled: bool) void {
+    const idx = @as(std.Target.Cpu.Feature.Set.Index, @intFromEnum(feature));
+
+    if (enabled) cpu.features.addFeature(idx) else cpu.features.removeFeature(idx);
+}
+
 pub fn detectNativeCpuAndFeatures(
     arch: std.Target.Cpu.Arch,
     os: std.Target.Os,
@@ -9,9 +19,6 @@ pub fn detectNativeCpuAndFeatures(
     _ = os;
     _ = query;
 
-    // Clearly this code could do better in the future by actually querying specific CPU features
-    // with the cpucfg instruction like on x86. But with the small number of well-known LoongArch
-    // models that exist at the moment, simply checking the PRID is plenty.
     var cpu: std.Target.Cpu = .{
         .arch = arch,
         .model = switch (cpucfg(0) & 0xf000) {
@@ -23,6 +30,31 @@ pub fn detectNativeCpuAndFeatures(
     };
 
     cpu.features.addFeatureSet(cpu.model.features);
+
+    const cfg1 = cpucfg(1);
+    const cfg2 = cpucfg(2);
+    const cfg3 = cpucfg(3);
+
+    setFeature(&cpu, .ual, bit(cfg1, 20));
+
+    const has_fpu = bit(cfg2, 0);
+    setFeature(&cpu, .f, has_fpu and bit(cfg2, 1));
+    setFeature(&cpu, .d, has_fpu and bit(cfg2, 2));
+
+    setFeature(&cpu, .lsx, bit(cfg2, 6));
+    setFeature(&cpu, .lasx, bit(cfg2, 7));
+    setFeature(&cpu, .lvz, bit(cfg2, 10));
+
+    setFeature(&cpu, .lbt, bit(cfg2, 18) and bit(cfg2, 19) and bit(cfg2, 20));
+
+    setFeature(&cpu, .frecipe, bit(cfg2, 25));
+    setFeature(&cpu, .div32, bit(cfg2, 26));
+    setFeature(&cpu, .lam_bh, bit(cfg2, 27));
+    setFeature(&cpu, .lamcas, bit(cfg2, 28));
+    setFeature(&cpu, .scq, bit(cfg2, 30));
+
+    setFeature(&cpu, .ld_seq_sa, bit(cfg3, 23));
+
     cpu.features.populateDependencies(cpu.arch.allFeaturesList());
 
     return cpu;
-- 
2.52.0

